src/experimental/tools/videoThumbnailLoader/thumbnail_loader.ts

```diff
@@ -0,0 +1,261 @@
+ * from a video source.
+ *
+ * The tools will extract a "thumbnail track" either from a video track (whose light
+ * chunks are adapted from such use case) or direclty from the media content.
+ */
+export default class VideoThumbnailLoader {
+  private readonly _thumbnailVideoElement: HTMLVideoElement;
+
+  private _manifest: Manifest;
+  private _currentJob?: IJob;
+
+  constructor(videoElement: HTMLVideoElement,
+              manifest: Manifest) {
+    this._thumbnailVideoElement = videoElement;
+    this._manifest = manifest;
+  }
+
+  /**
+   * Set time of thumbnail video media element :
+   * - Remove buffer when too much buffered data
+   * - Search for thumbnail track element to display
+   * - Load data
+   * - Append data
+   * Resolves when time is set.
+   * @param {number} time
+   * @returns {Promise}
+   */
+  setTime(time: number): Promise<unknown> {
+    return new PPromise((done, failed) => {
+      this._setTime({ time, done, failed }, this._thumbnailVideoElement);
+    });
+  }
+
+  /**
+   * Dispose thumbnail loader.
+   * @returns {void}
+   */
+  dispose(): void {
+    disposeSourceBuffer();
+    return this._currentJob?.stop();
+  }
+
+  private _setTime = (payload: { time: number;
+                                 done: (time: number) => void;
+                                 failed: (err: unknown) => void; },
+                      videoElement: HTMLVideoElement
+  ): void => {
+    const { time, done, failed } = payload;
+    if (time === this._thumbnailVideoElement.currentTime) {
+      return done(time);
+    }
+    const period = this._manifest.getPeriodForTime(payload.time);
+    if (period === undefined ||
+        period.adaptations.video === undefined ||
+        period.adaptations.video.length === 0) {
+      return failed(new VideoThumbnailLoaderError("NO_TRACK",
+                                                  "Couldn't find track for this time."));
+    }
+    const videoAdaptation = period.adaptations.video[0];
+    const representation = videoAdaptation.trickModeTrack?.representations[0] ??
+                           videoAdaptation.representations[0];
+    if (representation === undefined) {
+      return failed(new VideoThumbnailLoaderError("NO_TRACK",
+                                                  "Couldn't find track for this time."));
+    }
+    const thumbnailTrack = getThumbnailTrack(representation);
+    if (thumbnailTrack.initURL === "") {
+      return failed(new VideoThumbnailLoaderError("NO_INIT_DATA", "No init data for track."));
+    }
+    const thumbnails = getWantedThumbnails(thumbnailTrack,
+                                           payload.time,
+                                           videoElement.buffered);
+    if (thumbnails === null) {
+      return failed(new VideoThumbnailLoaderError("NO_THUMBNAILS", "Couldn't find thumbnail."));
+    }
+    if (thumbnails.length === 0) {
+      log.debug("VTL: Thumbnail already loaded.");
```

Comment:
We could also have no segment at all, no?

Reply:
Done